File: src/main/java/org/bbrk24/amurians/EmeryTableRecipeSerializer.java
package org.bbrk24.amurians;

import com.google.gson.JsonObject;

import net.minecraft.item.ItemStack;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.recipe.Ingredient;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.util.Identifier;
import net.minecraft.util.JsonHelper;
import net.minecraft.util.registry.Registry;

// pretty much just copied from CuttingRecipe$Serializer, since its constructor is inaccessible here
public class EmeryTableRecipeSerializer implements RecipeSerializer<EmeryTableRecipe> {
    @Override
    public EmeryTableRecipe read(Identifier identifier, JsonObject jsonObject) {
        String string = JsonHelper.getString(jsonObject, "group", "");
        Ingredient ingredient =
            JsonHelper.hasArray(jsonObject, "ingredient")
                ? Ingredient.fromJson(JsonHelper.getArray(jsonObject, "ingredient"))
                : Ingredient.fromJson(JsonHelper.getObject(jsonObject, "ingredient"));
        String string2 = JsonHelper.getString(jsonObject, "result");
        int i = JsonHelper.getInt(jsonObject, "count");
        ItemStack itemStack = new ItemStack(Registry.ITEM.get(new Identifier(string2)), i);
        return new EmeryTableRecipe(identifier, string, ingredient, itemStack);
    }

    @Override
    public EmeryTableRecipe read(Identifier identifier, PacketByteBuf packetByteBuf) {
        String string = packetByteBuf.readString();
        Ingredient ingredient = Ingredient.fromPacket(packetByteBuf);
        ItemStack itemStack = packetByteBuf.readItemStack();
        return new EmeryTableRecipe(identifier, string, ingredient, itemStack);
    }

    @Override
    public void write(PacketByteBuf packetByteBuf, EmeryTableRecipe cuttingRecipe) {
        packetByteBuf.writeString(cuttingRecipe.getGroup());
        cuttingRecipe.getInput().write(packetByteBuf);
        packetByteBuf.writeItemStack(cuttingRecipe.getOutput());
    }
}

File: src/main/kotlin/org/bbrk24/amurians/EmeryTableRecipeSerializer.kt
package org.bbrk24.amurians

import com.google.gson.JsonObject

import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraft.recipe.Ingredient
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.util.Identifier
import net.minecraft.util.JsonHelper
import net.minecraft.util.registry.Registry

object EmeryTableRecipeSerializer : RecipeSerializer<EmeryTableRecipe> {
    override fun read(id: Identifier, json: JsonObject): EmeryTableRecipe {
        // Kotlin can't parse the @Contract annotation, but this will never return null if the third
        // argument is non-null.
        val group = JsonHelper.getString(json, "group", "")!!

        val ingredient = Ingredient.fromJson(
            if (JsonHelper.hasArray(json, "ingredient"))
                JsonHelper.getArray(json, "ingredient")
            else
                JsonHelper.getObject(json, "ingredient")
        )
        val resultId = Identifier(JsonHelper.getString(json, "result"))
        val resultCount = JsonHelper.getInt(json, "count", 1)
        return EmeryTableRecipe(
            id,
            group,
            ingredient,
            ItemStack(Registry.ITEM.get(resultId), resultCount)
        )
    }

    override fun read(id: Identifier, buf: PacketByteBuf): EmeryTableRecipe {
        val group = buf.readString()
        val ingredient = Ingredient.fromPacket(buf)
        val output = buf.readItemStack()
        return EmeryTableRecipe(id, group, ingredient, output)
    }

    override fun write(buf: PacketByteBuf, recipe: EmeryTableRecipe) {
        buf.writeString(recipe.getGroup())
        recipe.getInput().write(buf)
        buf.writeItemStack(recipe.getOutput())
    }
}
